### DISCUSSION TITLE
Using a JSON column in comparisons?

### DISCUSSION THREAD

**Discussion by pallyndr** [moj-analytical-services/splink#2473] [Q&A]
Hello,

My use case is a DataFrame, with a "person" column, that contains JSON dictionary.
I would like to compare on personName.
I tried syntax like person->>'personName' but that does not work.

Is that supported?
If yes, what is the syntax?

Thanks!

**Comment by ADBond** [comment]
This will depend on the SQL backend you are using, and how it represents JSON or JSON-like types.

For instance, in DuckDB the equivalent type would be a `struct`, and you can [access fields with a `column.attribute` syntax](https://duckdb.org/docs/sql/data_types/struct#retrieving-from-structs). For another backend you would have to check the docs on that SQL dialect.

You should be able to use the expression `person.personName` in your comparisons where you may normally just have a single column, but it might be easier to make a derived table where you extract these attributes first, and use that as input to Splink, e.g.:
```python
import duckdb
import pandas as pd

df = pd.DataFrame(
    [
        {"unique_id": i,
        "person": {"personName": f"name_{i}", "age": 3*i + 2}}
        for i in range(20)
    ]
)

con = duckdb.connect()

con.sql(
"""
CREATE TABLE flattened_input AS
(SELECT *, person.personName AS name, person.age AS age FROM df")
"""
)
con.table("flattened_input").show()
# with this input table you can now refer simply to columns 'name', 'age' in your settings

...
linker = Linker("flattened_input", settings, DuckDBAPI(con))

```

**Reply by pallyndr** [reply]
My backend is Spark.

I agree plan B is create an intermediate table when the JSON attributes are "exploded" into individual columns.
I would like to avoid this.

person.personName does not work for Spark, gives some syntax error in SQL.
Also person->>'personName" fails.

Any idea here?

**Reply by ADBond** [reply]
The `person.personName` syntax should work in spark if your `person` column is a [struct](https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.types.StructType.html). If it is a string column then you can convert it to a struct with the [`from_json` function](https://spark.apache.org/docs/3.5.3/api/sql/index.html#from_json).

```python
from pyspark.sql import SparkSession
import pandas as pd

spark = SparkSession.builder.getOrCreate()

df = pd.DataFrame(
    [
        {"unique_id": i,
        "person": {"personName": f"name_{i}", "age": 3*i + 2}}
        for i in range(20)
    ]
)

spark_df = spark.createDataFrame(df)
spark_df.printSchema()

spark_df.createOrReplaceTempView("input")

spark.sql("""
SELECT *, person.personName AS name, person.age AS age FROM input
"""
).show()
```
